tests/src/junit/java/net/neoforged/neoforge/unittest/transfer/InstancedHandlerTests.java

```diff
@@ -0,0 +1,115 @@
+import org.assertj.core.api.ThrowableAssert;
+import org.junit.jupiter.api.MethodOrderer;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestMethodOrder;
+
+@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
+public class InstancedHandlerTests {
+    @Test
+    void emptyHandlers() {
+        //EMPTY no operation handlers but should throw on index calls
+        testEmptyHandler(EmptyResourceHandler.instance(), ItemResource.EMPTY);
+        testEmptyHandler(EmptyResourceHandler.instance(), FluidResource.EMPTY);
+    }
+
+    @Test
+    public void voidHandlers() {
+        //VoidResourceHandlers destroys resources but doesn't allow extraction
+        testVoidResource(VoidResourceHandler.ITEM, ItemResource.EMPTY);
+        testVoidResource(VoidResourceHandler.FLUID, FluidResource.EMPTY);
+    }
+
+    @Test
+    public void endlessHandlers() {
+        //InfiniteResourceHandlers creates infinite of a specified resource, but doesn't allow insertion
+        testEndlessResource(FluidResource.of(Fluids.WATER), FluidResource.EMPTY);
+        testEndlessResource(ItemResource.of(Blocks.COBBLESTONE), ItemResource.EMPTY);
+    }
+
+    private static <T extends IResource> void testVoidResource(VoidResourceHandler<T> handler, T emptyResource) {
+        Assertions.assertThat(handler.size()).withFailMessage("Size should be 1").isEqualTo(1);
+        Assertions.assertThat(handler.supportsExtraction()).withFailMessage("Extraction should be not be allowed").isFalse().isEqualTo(handler.supportsExtraction(0));
+        Assertions.assertThat(handler.supportsInsertion()).withFailMessage("Insertion should be be allowed").isTrue().isEqualTo(handler.supportsInsertion(0));
+        Assertions.assertThat(ResourceHandlerUtil.isValid(handler, emptyResource)).withFailMessage("Voids should always be valid for a resource").isTrue();
+        Assertions.assertThat(handler.getCapacity(0, emptyResource)).withFailMessage("Voids should hava capacity of max int").isEqualTo(Integer.MAX_VALUE);
+        Assertions.assertThat(handler.getResource(0)).withFailMessage("Empty resource should match").isEqualTo(emptyResource);
+
+        try (var transaction = TransactionManager.open(null)) {
+            Assertions.assertThat(Integer.MAX_VALUE)
+                    .withFailMessage("Void should be able to accept infinite resources")
+                    .isEqualTo(handler.insert(0, emptyResource, Integer.MAX_VALUE, transaction))
+                    .isEqualTo(handler.insert(emptyResource, Integer.MAX_VALUE, transaction));
+
+            Assertions.assertThat(0)
+                    .withFailMessage("Void should be provide no resources")
+                    .isEqualTo(handler.extract(0, emptyResource, Integer.MAX_VALUE, transaction))
+                    .isEqualTo(handler.extract(emptyResource, Integer.MAX_VALUE, transaction));
+        }
+    }
+
+    private static <T extends IResource> void testEndlessResource(T resource, T emptyResource) {
```

Comment:
`Infinite` is a better name than `Endless`, and matches the name of the actual handlers